go/nbs/table_persister.go

```diff
@@ -4,7 +4,94 @@
 
 package nbs
 
+import (
+	"bytes"
+	"sync"
+
+	"github.com/attic-labs/noms/go/d"
+	"github.com/attic-labs/noms/go/util/sizecache"
+)
+
 type tablePersister interface {
 	Compact(mt *memTable, haver chunkReader) chunkSource
+	CompactAll(sources chunkSources) chunkSource
 	Open(name addr, chunkCount uint32) chunkSource
 }
+
+type indexCache struct {
+	cache *sizecache.SizeCache
+}
+
+// Returns an indexCache which will burn roughly |size| bytes of memory
+func newIndexCache(size uint64) *indexCache {
+	return &indexCache{sizecache.New(size)}
+}
+
+func (sic indexCache) get(name addr) (tableIndex, bool) {
+	idx, found := sic.cache.Get(name)
+	if found {
+		return idx.(tableIndex), true
+	}
+
+	return tableIndex{}, false
+}
+
+func (sic indexCache) put(name addr, idx tableIndex) {
+	indexSize := uint64(idx.chunkCount) * (addrSize + ordinalSize + lengthSize + uint64Size)
+	sic.cache.Add(name, indexSize, idx)
+}
+
+type chunkSourcesByDescendingCount chunkSources
+
+func (csbc chunkSourcesByDescendingCount) Len() int { return len(csbc) }
+func (csbc chunkSourcesByDescendingCount) Less(i, j int) bool {
+	srcI, srcJ := csbc[i], csbc[j]
+	if srcI.count() == srcJ.count() {
+		hi, hj := srcI.hash(), srcJ.hash()
+		return bytes.Compare(hi[:], hj[:]) > 0
+	}
+	return srcI.count() > srcJ.count()
+}
+func (csbc chunkSourcesByDescendingCount) Swap(i, j int) { csbc[i], csbc[j] = csbc[j], csbc[i] }
+
+func compactSourcesToBuffer(sources chunkSources, rl chan struct{}) (name addr, data []byte, chunkCount uint32) {
+	d.Chk.True(rl != nil)
+	totalData := uint64(0)
+	for _, src := range sources {
+		chunkCount += src.count()
+		totalData += src.data()
+	}
+	if chunkCount == 0 {
+		return
+	}
+
+	maxSize := maxTableSize(uint64(chunkCount), totalData)
+	buff := make([]byte, maxSize) // This can blow up RAM (BUG 3130)
+	tw := newTableWriter(buff)
+
+	// Use "channel of channels" ordered-concurrency pattern so that chunks from a given table stay together, preserving whatever locality was present in that table.
+	chunkChans := make(chan chan extractRecord)
```

Comment:
I may be understanding the question wrong, but yes, it will load all input tables concurrently (modulo the rate-limiter channel `rl`). We talked about this the other day, and when I told you that it took a while to do a compaction, you pointed this out as an optimization opportinity.

I used the "channel of channels" pattern that I picked up from your readahead to ensure that the chunks from the different tables don't get all mixed up together. That's what the comment above is referring to.